File: Leetcode/Typescript/3335. Total Characters in String After Transformations I/solution.ts
// https://leetcode.com/problems/total-characters-in-string-after-transformations-i/description

function lengthAfterTransformations(s: string, t: number): number {
  const MOD = 1_000_000_007;
  const queue: number[] = new Array(26).fill(0);

  for (const char of s) {
    const idx = char.charCodeAt(0) - "a".charCodeAt(0);
    queue[idx]++;
  }

  for (let step = 0; step < t; step++) {
    const z = queue[25] % MOD;
    for (let i = 25; i > 0; i--) {
      queue[i] = queue[i - 1];
    }
    queue[0] = z;
    queue[1] = (queue[1] + z) % MOD;
  }

  return queue.reduce((acc, val) => (acc + val) % MOD, 0);
}

File: Leetcode/Python/3335. Total Characters in String After Transformations I/solution.py
# https://leetcode.com/problems/total-characters-in-string-after-transformations-i/description

# Rotating queue, 330 ms
from collections import deque


MOD = 10 ** 9 + 7
A_CODE = ord('a')
class Solution:
    def lengthAfterTransformations(self, s: str, t: int) -> int:
        queue = deque([0] * 26)

        # Character counter
        for char in s:
            code = ord(char) - A_CODE
            queue[code] += 1
        
        # Transform 'z' character
        for _ in range(t):
            count = queue.pop() % MOD

            queue.appendleft(count)
            queue[1] = (queue[1] + count) % MOD

        return sum(queue) % MOD


# Implement, 2300 ms
MOD = 10 ** 9 + 7
A_CODE = ord('a')

class Solution:
    def lengthAfterTransformations(self, s: str, t: int) -> int:
        count = [0] * 26

        for c in s:
            count[ord(c) - A_CODE] += 1

        for _ in range(t):
            newCount = [0] * 26
            for i in range(25):
                newCount[i + 1] = count[i]
            newCount[0] = count[25]
            newCount[1] = (newCount[1] + count[25]) % MOD
            count = newCount

        return sum(count) % MOD
    